data/cell.hh

```diff
@@ -261,7 +259,7 @@ struct cell {
                 }
             }
             template<typename Tag>
-            size_t size_of() const noexcept {
+            size_t size_of(Tag) const noexcept {
```

Comment:
We should de-template these methods. I think you can just remove the template and it will keep working.

Reply:
As I wrote, this is a trivial transfromation.
But, if I remove the template I will need a size_of(Tag) for every tag that tags a buffer.
Meaning I will need to have separate implementations for:
```
size_t size_of(T1) {
 return ...;
}

size_t size_of(T2) {
return ...;
}
...
```
This is what you wanted in the first place didn't you?

BTW, in the current implementation there is exactly one specialization for this method, meaning that even if I did this separation I will end up with a bunch of identical overloaded functions and one more which is different.
But so I am sure that I am not talking complete nonsense I will remove the template and see if it compile... :smile:

Reply:
@denesb  It didn't work, I got a lot of errors of this style:
```
...
error: no matching function for call to 'imr::buffer<data::cell::tags::data>::serialized_object_size(const uint8_t*&, data::cell::variable_value::context)'
  245 |             return alternative::type::serialized_object_size(in, context.template context_for<typename alternative::tag>(in));
      |                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from ./data/cell.hh:30,
                 from ./types.hh:27,
                 from ./dht/i_partitioner.hh:45,
                 from ./dht/sharder.hh:24,
                 from sstables/sstables.cc:39:
./imr/fundamental.hh:260:19: note: candidate: 'static size_t imr::buffer<Tag>::serialized_object_size(const uint8_t*, const Context&) [with Context = data::cell::variable_value::context; Tag = data::cell::tags::data; size_t = long unsigned int; uint8_t = unsigned char]'
  260 |     static size_t serialized_object_size(const uint8_t*, const Context& context) noexcept {
```
Basically it boils down to that:
The compiler searched for a function `size_t size_of(data::cell::tags::data)` in `data::cell::variable_value::context` and couldn't find it. (which in turn didn't generate the `serialized_object_size` function which calls size_of).
We will get this error at least for every concrete tag that is used in a buffer for the corresponding context.

With all of that said, it is your project and if you think this is the way it should be I will do it.
I can also do it in a follow up step and then we will consider if to get this follow up in or not.

Reply:
@denesb ping, what do you decide? Should I do it in this step or should I do it in a follow-up?

Reply:
The point of making these functions overloads is to remove the template, because leaving it as a template leaves the possibility of mistakes open. A new buffer can be added, or one changed and it will happily keep using the out-of-date `size_of()` in the context.

> But, if I remove the template I will need a size_of(Tag) for every tag that tags a buffer.

Yes, that would be the point. But don't we already have one of these per buffer already? Just as a template specialization, not an overload?

> Meaning I will need to have separate implementations for:
>```
> size_t size_of(T1) {
> return ...;
> }
>
>size_t size_of(T2) {
> return ...;
> }
> ...
>```
>
> This is what you wanted in the first place didn't you?

Yes. I'm fine with this being done in a follow-up as long as it's going to happen.